### PR TITLE
THREESCALE-10843: Audit user sessions

### PR DESCRIPTION
**Pull request by jlledom**
**What this PR does / why we need it**:

This PR adds audits for `UserSession`. In particular fot `:create` and `:revoke` operations

The Jira issue also mentions auditing when a `User` is locked. I'm not sure what **"locked"** means but I assume it means **"suspended"**. If that's the case, then that was already done.

Besides, the issue also asks to add some audit logs for some operations:

- Login
  - success (this was already done)
  - failed
- User account
  - password change success
  - password change failed

In both admin and developer portals.

The PR also adds some tests for the new functionality.

**Which issue(s) this PR fixes** 

https://issues.redhat.com/browse/THREESCALE-10843

**Verification steps** 

Just perform the operations above and check the audits appear.

### CODE REVIEW

File: app/controllers/provider/admin/user/personal_details_controller.rb [hunk 1]
```diff
@@ -34,6 +35,7 @@ def current_password_verification
 
     unless current_user.authenticated?(user_params[:current_password])
       flash.now[:error] = 'Current password is incorrect.'
+      AuditLogService.call("User tried to change password: #{current_user.id}/#{current_user.username}") if user_params[:password].present?
```
Comment (mayorova) [inline app/controllers/provider/admin/user/personal_details_controller.rb:38]: In this case, I'd add "tried, but failed due to incorrect password". Because it's clear that they tried, but not clear if they succeeded.

Comment (jlledom) [reply app/controllers/provider/admin/user/personal_details_controller.rb]: Done here: https://github.com/3scale/porta/pull/3885/commits/3279c2334d276b700340b4c528937d41f70e4c4b

File: app/controllers/provider/admin/user/personal_details_controller.rb [hunk 2]
```diff
@@ -34,6 +35,7 @@ def current_password_verification
 
     unless current_user.authenticated?(user_params[:current_password])
       flash.now[:error] = 'Current password is incorrect.'
+      AuditLogService.call("User tried to change password, but failed due to incorrect current password: #{current_user.id}/#{current_user.username}") if user_params[:password].present?
```
Comment (akostadinov) [inline app/controllers/provider/admin/user/personal_details_controller.rb:38]: why `user_params[:password].present?` ?

File: app/controllers/provider/sessions_controller.rb [hunk 3]
```diff
@@ -34,6 +34,8 @@ def create
     else
       new
       flash.now[:error] ||= strategy&.error_message
+      attempted_cred = auth_params.fetch(:username, nil) || sso_params.fetch(:token, nil)
+      AuditLogService.call("Login attempt failed: #{domain_account.external_admin_domain} - #{attempted_cred}")
```
Comment (akostadinov) [inline app/controllers/provider/sessions_controller.rb:38]: is token safe to log? Can it be a valid token for something else so putting in logs to be inappropriate? What do we already have in `strategy&.error_message`? I wonder if we can take example from it.

File: lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb [hunk 4]
```diff
@@ -19,7 +19,10 @@ def show
   def update
     resource.validate_fields!
     if resource.errors.empty? && resource.update(user_params)
-      resource.kill_user_sessions(user_session) if resource.just_changed_password?
+      if resource.just_changed_password?
+        AuditLogService.call("User changed password: #{resource.id}/#{resource.username}")
```
Comment (akostadinov) [inline lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb:23]: Don't we audit the whatever object stores the password already?

File: lib/developer_portal/app/controllers/developer_portal/login_controller.rb [hunk 5]
```diff
@@ -31,19 +31,29 @@ def create
       self.current_user = @user
       create_user_session!
       flash[:notice] = @strategy.new_user_created? ? 'Signed up successfully' : 'Signed in successfully'
+
+      AuditLogService.call("Signed in: #{current_user.id}/#{current_user.username} #{current_user.first_name} #{current_user.last_name}")
```
Comment (akostadinov) [inline lib/developer_portal/app/controllers/developer_portal/login_controller.rb:35]: Isn't full name a little bit too much logging?

Also, provided that we audit UserSession, why do we need a separate log message that a user logged in? Maybe one can watch the sessions creation audits.

File: lib/developer_portal/app/controllers/developer_portal/login_controller.rb [hunk 6]
```diff
@@ -31,19 +31,29 @@ def create
...
+
       redirect_back_or_default(@strategy.redirect_to_on_successful_login)
     elsif @strategy.redirects_to_signup?
       @strategy.on_signup(session)
+
       redirect_to @strategy.signup_path(params), notice: 'Successfully authenticated, please complete the signup form'
     else
+      attempted_cred = auth_strategy_is_internal? ? params.fetch(:username, nil) : params.fetch(:token, nil)
+      AuditLogService.call("Login attempt failed: #{request.internal_host} - #{attempted_cred}")
```
Comment (akostadinov) [inline lib/developer_portal/app/controllers/developer_portal/login_controller.rb:44]: same question about token here

File: lib/developer_portal/app/controllers/developer_portal/login_controller.rb [hunk 7]
```diff
@@ -31,19 +31,29 @@ def create
...
       render_login_error(@strategy.error_message)
     end
   end
 
   def destroy
+    user = current_user
     logout_killing_session!
     destroy_user_session!
 
+    AuditLogService.call("Signed out: #{user.id}/#{user.username} #{user.first_name} #{user.last_name}")
```
Comment (akostadinov) [inline lib/developer_portal/app/controllers/developer_portal/login_controller.rb:55]: Similar here, we audit UserSession. Maybe no need to also log that a user signed out.

File: app/controllers/provider/sessions_controller.rb (hunk 3 above)
Comment (jlledom) [reply app/controllers/provider/sessions_controller.rb:38]: I didn't think on this. Maybe we should not log the token. I'd say those tokens are one timers with a short expiration time. But not 100% sure so better remove it, I think. We can log the attempted username if any, if a SSO login failed, then we don't log the user because we don't have it.

File: lib/developer_portal/app/controllers/developer_portal/login_controller.rb (hunk 5 above)
Comment (jlledom) [reply lib/developer_portal/app/controllers/developer_portal/login_controller.rb:35]: > Isn't full name a little bit too much logging?

Maybe, I copied this from what we do in the admin portal:

https://github.com/3scale/porta/blob/ca21bb0a1fa8f6dd1a8a4773ad0891f5e8c666b1/app/controllers/provider/sessions_controller.rb#L31

> Also, provided that we audit UserSession, why do we need a separate log message that a user logged in? Maybe one can watch the sessions creation audits.

Maybe it's not required. I added the log because that's what the issue description says. But I'm open to remove this traces if you think they are not needed.

File: app/controllers/provider/sessions_controller.rb (hunk 3 above)
Comment (jlledom) [reply app/controllers/provider/sessions_controller.rb:38]: Done here: https://github.com/3scale/porta/pull/3885/commits/38d22eada58fcf5802a585724ac15872c82df170

File: app/controllers/provider/admin/user/personal_details_controller.rb (hunk 2 above)
Comment (akostadinov) [reply app/controllers/provider/admin/user/personal_details_controller.rb:38]: should we rather check the future password fields instead? Honest question. Not sure that would be better.

File: lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb (hunk 4 above)
Comment (akostadinov) [reply lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb:23]: I think it makes sense to start auditing that field. Just redact the value.

https://github.com/collectiveidea/audited/blob/9aad4d46e52625d1608917df22d0f31fbd873db4/lib/audited/auditor.rb#L40-L48

File: app/controllers/provider/sessions_controller.rb (hunk 3 above)
Comment (akostadinov) [reply app/controllers/provider/sessions_controller.rb:38]: Maybe log this `strategy&.error_message` appended to the existing one. Also will be very useful to log the client IP address I believe. Will be helpful before resorting to traffic capture.

File: app/controllers/provider/admin/user/personal_details_controller.rb (hunk 2 above)
Comment (akostadinov) [reply app/controllers/provider/admin/user/personal_details_controller.rb:38]: To me it seems like if `password` is set, then user tried to change password, and used wrong password or no password (which i technically also a wrong password).

If only current password was set but not new password, then user didn't try to change password. They just added noise.

Perhaps not a big deal but the intention IMO is shown by filling in the new password.

File: lib/developer_portal/app/controllers/developer_portal/login_controller.rb (hunk 5 above)
Comment (jlledom) [reply lib/developer_portal/app/controllers/developer_portal/login_controller.rb:35]: And what about the audit log already existing in the admin portal, should we remove it also? It's not required anymore now we are auditing `UserSession`

File: app/controllers/provider/sessions_controller.rb (hunk 3 above)
Comment (jlledom) [reply app/controllers/provider/sessions_controller.rb:38]: Yeah I can do that 👍

File: app/controllers/provider/admin/user/personal_details_controller.rb (hunk 2 above)
Comment (jlledom) [reply app/controllers/provider/admin/user/personal_details_controller.rb:38]: I think we agree, and the code does what you say. Don't you think so?

File: lib/developer_portal/app/controllers/developer_portal/login_controller.rb (hunk 5 above)
Comment (akostadinov) [reply lib/developer_portal/app/controllers/developer_portal/login_controller.rb:35]: +1, with a note in the release notes.

Comment (akostadinov) [reply lib/developer_portal/app/controllers/developer_portal/login_controller.rb:35]: Aternatively we may stop auditing session and keep only the log :sunglasses:

File: app/controllers/provider/admin/user/personal_details_controller.rb (hunk 2 above)
Comment (akostadinov) [reply app/controllers/provider/admin/user/personal_details_controller.rb:38]: My complained was actually about
```
unless current_user.authenticated?(user_params[:current_password])
```

Why flash a message if user never supplied `:password` parameter? They didn't try to change password so they don't care that it didn't happen.

Comment (jlledom) [reply app/controllers/provider/admin/user/personal_details_controller.rb:38]: Because if they don't provide `current_password` they can't change anything, not only the `password` field.

Comment (akostadinov) [reply app/controllers/provider/admin/user/personal_details_controller.rb:38]: got it, now it makes sense

File: app/controllers/provider/sessions_controller.rb [hunk 8]
```diff
@@ -28,12 +28,12 @@ def create
       create_user_session!(strategy&.authentication_provider_id)
       flash[:notice] = 'Signed in successfully'
 
-      AuditLogService.call("Signed in: #{current_user.id}/#{current_user.username} #{current_user.first_name} #{current_user.last_name}")
```
Comment (mayorova) [inline app/controllers/provider/sessions_controller.rb:31]: Is this log removed because now there are audit logs for the "UserSession" model?

Comment (jlledom) [reply app/controllers/provider/sessions_controller.rb:31]: Yes, exactly. Now we audit `UserSession` and also `User#password_digest`, so we don't need augit logs for login/logout and user password change

File: app/controllers/provider/sessions_controller.rb [hunk 9]
```diff
@@ -28,12 +28,12 @@ def create
...
-
       redirect_back_or_default provider_admin_path
     else
       new
       flash.now[:error] ||= strategy&.error_message
+      attempted_cred = auth_params.fetch(:username, 'SSO')
```
Comment (mayorova) [inline app/controllers/provider/sessions_controller.rb:35]: Why the fallback is "SSO"? :thinking:

File: app/controllers/provider/sessions_controller.rb (hunk 3 above)
Comment (jlledom) [reply app/controllers/provider/sessions_controller.rb]: Done: https://github.com/3scale/porta/pull/3885/commits/22ce1889708acf065aa65f3bf90999d68794e218

File: app/controllers/provider/sessions_controller.rb (hunk 9 above)
Comment (mayorova) [reply app/controllers/provider/sessions_controller.rb:35]: hm, I am not quite sure about the SSO flow, I'd need to check, but would there potentially be some value in logging the `system_name` of the auth provider instead?... just thinking out loud.

File: lib/developer_portal/app/controllers/developer_portal/login_controller.rb (hunk 5 above)
Comment (jlledom) [reply lib/developer_portal/app/controllers/developer_portal/login_controller.rb]: Also added some release notes to the issue.

File: app/controllers/provider/sessions_controller.rb (hunk 9 above)
Comment (jlledom) [reply app/controllers/provider/sessions_controller.rb:35]: Maybe, that would allow admins to know which auth provider was being used when it failed to log in. Probably that's better than just print `SSO`

File: lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb [hunk 10]
```diff
@@ -19,7 +19,9 @@ def show
   def update
     resource.validate_fields!
     if resource.errors.empty? && resource.update(user_params)
-      resource.kill_user_sessions(user_session) if resource.just_changed_password?
+      if resource.just_changed_password?
+        resource.kill_user_sessions(user_session)
+      end
```
Comment (akostadinov) [inline lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb:24]: why this change? Seems like rubocop prefers single line :)

Comment (akostadinov) [reply lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb:24]: But actually wasn't `password_digest` attr important now?

**Review by akostadinov — APPROVED** [review]
I have a line question that is probably irrelevant. And would suggest adding a test for the password change audit unless I missed one existing.

Approved in case its just my oversight.

File: lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb (hunk 10 above)
Comment (jlledom) [reply lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb:24]: > why this change? Seems like rubocop prefers single line :)

Yeah, it was single line before but I made it complete form because I was also adding the audit log here. But then I removed the audit log and forgot to make it single line again.

> But actually wasn't password_digest attr important now?

`password` is a field in the form, not the DB column, that is called `password_digest` since we don't save passwords in plain text. In fact, look how is `just_changed_password` implemented:

https://github.com/3scale/porta/blob/843f799f292f34cb6e26fa5e9ab2a34526ef2ae3/app/lib/authentication/by_password.rb#L41-L43

Comment (jlledom) [reply lib/developer_portal/app/controllers/developer_portal/admin/account/personal_details_controller.rb]: Fixed the `if`: https://github.com/3scale/porta/pull/3885/commits/df2fa1986850eca6f32cb5720921b13c14f2d67d

File: test/functional/developer_portal/admin/account/personal_details_controller_test.rb [hunk 11]
```diff
@@ -63,4 +63,19 @@ def setup
     assert_response :success
     assert_equal flash[:error], 'Current password is incorrect'
   end
+
+  test 'changing password is audited' do
+    login_as @buyer.admins.first
+    assert_difference(Audited.audit_class.method(:count)) do
+      User.with_synchronous_auditing do
+        put :update, params: { user: {current_password: 'supersecret', password: 'new_password', password_confirmation: 'new_password'} }
```
Comment (akostadinov) [inline test/functional/developer_portal/admin/account/personal_details_controller_test.rb:71]: also asserting that the value of old and new password in audit is `[REDACTED]` could be useful

**Comment by jlledom** [comment]
> And would suggest adding a test for the password change audit unless I missed one existing.

I added that test for provider and developer portals: https://github.com/3scale/porta/pull/3885/commits/7607ae06e5cf07af5105bf4bbea22e0a80f98ba8

So please, approve or die.

File: test/functional/developer_portal/admin/account/personal_details_controller_test.rb (hunk 11 above)
Comment (jlledom) [reply test/functional/developer_portal/admin/account/personal_details_controller_test.rb:71]: Mmmm, actually...

```ruby
[27] pry(main)> user.audits.last
=> #<Audited::Audit:0x000000004247d5a0
 id: 1454,
 auditable_id: 2,
 auditable_type: "User",
 user_id: 2,
 user_type: "User",
 username: nil,
 action: "update",
 version: 2,
 created_at: Thu, 19 Sep 2024 12:02:25.000000000 UTC +00:00,
 tenant_id: 2,
 provider_id: 2,
 kind: "User",
 audited_changes: {"password_digest"=>"[FILTERED]"},
 comment: nil,
 associated_id: nil,
 associated_type: nil,
 remote_address: "127.0.0.1",
 request_uuid: "d97ee5f5-ccf1-4dc8-a39a-e4e2aec767e3">

[28] pry(main)> user.audits.last.audited_changes
=> {"password_digest"=>
  ["$2a$12$TshOAone/WSVbXsCmOB1JuLGBQI698CButPpA2F7hw1jEb2db4YWi",
   "$2a$12$.0G4VigmmeE1cjozQbXPkuyYF5iBjcfIwtjqcqJCPomceKUr7REkS"]}
```

Is that expected?

Comment (jlledom) [reply test/functional/developer_portal/admin/account/personal_details_controller_test.rb:71]: Shit. It was actually not redacting the value. I'll fix it and add the test.

Comment (jlledom) [reply test/functional/developer_portal/admin/account/personal_details_controller_test.rb:73]: Fixed: https://github.com/3scale/porta/pull/3885/commits/efa0fdf398c33d5eca8c77a11b2cf5eff7c31b3f

And updated the tests: https://github.com/3scale/porta/pull/3885/commits/5f93b2c880c0eef43e8e627c4fae0e83dab432a3

File: test/functional/developer_portal/admin/account/personal_details_controller_test.rb [hunk 12]
```diff
@@ -65,12 +65,17 @@ def setup
...
+    login_as user
+
     assert_difference(Audited.audit_class.method(:count)) do
       User.with_synchronous_auditing do
         put :update, params: { user: {current_password: 'supersecret', password: 'new_password', password_confirmation: 'new_password'} }
       end
     end
+
+    expected = [Audited::Auditor::AuditedInstanceMethods::REDACTED] * 2
+    assert_equal expected,user.audits.last.audited_changes['password_digest']
```
Comment (akostadinov) [inline test/functional/developer_portal/admin/account/personal_details_controller_test.rb:78]: I don't understand why it is `*2` but better than the actual value :)

**Review by akostadinov — APPROVED** [review]

File: test/functional/developer_portal/admin/account/personal_details_controller_test.rb (hunk 12 above)
Comment (jlledom) [reply test/functional/developer_portal/admin/account/personal_details_controller_test.rb:78]:
```
[Audited::Auditor::AuditedInstanceMethods::REDACTED] * 2
```

is a cool way to do

```
[Audited::Auditor::AuditedInstanceMethods::REDACTED, Audited::Auditor::AuditedInstanceMethods::REDACTED]
```

Comment (akostadinov) [reply test/functional/developer_portal/admin/account/personal_details_controller_test.rb:78]: ah, now I saw the `[]`
